### PR TITLE
Issue #7733: Update AbstractChecks to log DetailAST - ImportOrder

### PR DESCRIPTION
**Pull request by DXTkastb**
Fixes issue #7733 

**Updates(total 5 log calls):**
https://github.com/checkstyle/checkstyle/blob/3953912f63bf368a5c586f276dca0d996cca3124/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L666

https://github.com/checkstyle/checkstyle/blob/3953912f63bf368a5c586f276dca0d996cca3124/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L673

https://github.com/checkstyle/checkstyle/blob/3953912f63bf368a5c586f276dca0d996cca3124/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L676

https://github.com/checkstyle/checkstyle/blob/3953912f63bf368a5c586f276dca0d996cca3124/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L738

https://github.com/checkstyle/checkstyle/blob/3953912f63bf368a5c586f276dca0d996cca3124/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java#L754

**diff report(default configuration):**
https://dxtkastb.github.io/checkstyle/reports/ImportOrder/diff/

**diff report for configuration:**
https://dxtkastb.github.io/checkstyle/reports/ImportOrder/diff2.0/

configuration:
```
 <property name="ordered" value="false"/>
 <property name="separated" value="true"/>
 <property name="option" value="above"/>
 <property name="sortStaticImportsAlphabetically" value="true"/>
```

### CODE REVIEW

**Comment by DXTkastb** [comment]
@rnveach I made the changes suggested by you, please review.
I had to close the other PR because I made some other changes and pushed them which made it a bit ambiguous. (old pr link: https://github.com/checkstyle/checkstyle/pull/7847).
Please let me know if more changes have to be made.

**Comment by baratali** [comment]
@DXTkastb , please fix travis-ci failure:
```
The PR is based on a master that is 13 commit(s) old.
The max allowed is 10.
This PR is too old and should be rebased on origin/master.
```

**Comment by DXTkastb** [comment]
@rnveach please review.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 1]
```diff
@@ -655,25 +654,27 @@ else if (option == ImportOrderOption.INFLOW) {
      * @param isStatic whether the token is static or not.
      * @param previous previous non-static but current is static (above), or
      *                  previous static but current is non-static (under).
-     * @param line the line of the current import.
+     * @param ast the line of the current import.
      */
-    private void doVisitToken(FullIdent ident, boolean isStatic, boolean previous, int line) {
+    private void doVisitToken(FullIdent ident, boolean isStatic, boolean previous, DetailAST ast) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: https://github.com/checkstyle/checkstyle/blob/ae9edbd4a8645c8cc1a99ad6890dd0006700d1cb/src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java#L44
`FullIdent` holds the AST too.

Is there any difference between the new AST passed in and the one in FullIdent? If so, what is the difference? If not, doesn't it make sense to use that instead?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 2]
```diff
@@ -655,25 +654,27 @@ else if (option == ImportOrderOption.INFLOW) {
      * @param isStatic whether the token is static or not.
      * @param previous previous non-static but current is static (above), or
      *                  previous static but current is non-static (under).
-     * @param line the line of the current import.
+     * @param ast the line of the current import.
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: If this stays, this isn't "the line" anymore.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 3]
```diff
@@ -728,14 +729,14 @@ private boolean isSeparatorBeforeImport(int line) {
      * @param previous previous non-static but current is static (above), or
      *     previous static but current is non-static (under).
      * @param name the name of the current import.
-     * @param line the line of the current import.
+     * @param ast the line of the current import.
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: Same.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 4]
```diff
@@ -973,5 +974,5 @@ else if (CommonUtil.startsWithChar(pkg, '/')) {
         }
         return patterns;
     }
-
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: Please restore this removed line. We don't expect any unnecessary changes.

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java [hunk 5]
```diff
@@ -667,5 +664,5 @@ protected void finishLocalSetup() throws CheckstyleException {
             }
         }
     }
-
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java]: same.

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java [hunk 6]
```diff
@@ -667,5 +664,5 @@ protected void finishLocalSetup() throws CheckstyleException {
...
 }
+
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java]: Please confirm this. Is this a second empty line? All Java files should have an empty line at the bottom already.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java [hunk 7]
```diff
@@ -930,5 +930,4 @@ public void testStaticGroupsUnordered() throws Exception {
         verify(checkConfig, getNonCompilablePath("InputImportOrderStaticGroupsUnordered.java"),
                 expected);
     }
-
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: same.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 1 above)
Comment (DXTkastb) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: @rnveach can you please elaborate me more on this change,

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: You changed the code to pass in the AST, but the AST is available from the `FullIdent`.

I am asking, are the 2 ASTs the same instance? If they are not, explain what each one points to.

Comment (DXTkastb) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: @rnveach I did a dry run for the code and found that the ast instance(present in ImportOrdercheck.java) traverses the nodes, whereas private DetailAST detailAst(present in the FullIdent.java) points specifically to the import package's first node name.
example:
import java.util.List;
ast:import>java>util>List(traversal)
detailAst:java
Code(in FullIdent.java):
``````
    private void append(DetailAST ast) {
        elements.add(ast.getText());
        if (detailAst == null) {
            detailAst = ast;
        }
    }
``````
this piece of code initializes detailAst to java(or whatever the first name be, say org/com...).
In terms of reference yes these two do point to the same instance but usability differs(pointing to a node). And I gave a try to your point as well, passing FullIdent.getDetailAst() rather than ast to the log can provide more precision as it throws error for import's line no. as well as the column no.(which was always '1'  due to imports alignment). It throws an error for incorrect expected column number:

``````
package org.checkstyle.suppressionxpathfilter.importorder;

import java.io.*;
import java.util.Set;
import static java.lang.Math.PI; // warn

public class SuppressionXpathRegressionImportOrderFour {
	// code
}
``````
``````
Expected :/home/kaustubh/main repo/checkstyle/src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderFour.java:5:1: Wrong order for 'java.lang.Math.PI' import.
Actual   :/home/kaustubh/main repo/checkstyle/src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderFour.java:5:15: Wrong order for 'java.lang.Math.PI' import.
 ``````
This happened because the passed value is now 'java' at 5:15: (FulIdent.detailAst) rather than 
 'import' at 5:1: (ast)

Should I change these passed values to ident.getDetailAst() or let them remain as ast?

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: > example:
import java.util.List;
ast:import>java>util>List(traversal)
detailAst:java

To confirm, `detailAst` is from FullIdent and `ast` is AST being passed in method parameter?
If so, will `detailAst,getParent()` always point to `ast` then?

Comment (DXTkastb) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: Yes, detailAst is from FullIdent and ast is the AST passed in method parameter of visitiToken();
No, detailAst.getParent() isn't pointing to ast.(They are pointing to different nodes)

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]:
````
IMPORT -> import [1:0]
|--DOT -> . [1:16]
|  |--DOT -> . [1:11]
|  |  |--IDENT -> java [1:7]
|  |  `--IDENT -> util [1:12]
|  `--IDENT -> List [1:17]
`--SEMI -> ; [1:21]
````

You are correct. I thought the tree looked different.

Ok, I am good to leave it as using the custom AST in the memory parameter. It makes more sense to have the violation be the import itself.

**Comment by rnveach** [comment]
Please resolve the other items mentioned.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 2 above)
Comment (DXTkastb) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 3 above)
Comment (DXTkastb) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: done. removed this line

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 4 above)
Comment (DXTkastb) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java (hunk 5 above)
Comment (DXTkastb) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java (hunk 6 above)
Comment (DXTkastb) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionXpathFilter.java]: fixed this too.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java (hunk 7 above)
Comment (DXTkastb) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheckTest.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 3 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: This line can't be removed. We have to document all method parameters.

Comment (DXTkastb) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java]: Ok I will document this, and will push the changes

**Comment by strkkk** [comment]
@rnveach there are unresolved items 
https://github.com/checkstyle/checkstyle/pull/7871#discussion_r393798221
https://github.com/checkstyle/checkstyle/pull/7871#discussion_r393798457

Please resolve them if it is ok

**Comment by rnveach** [comment]
@strkkk 2 admins have signed off on this PR. Since you are the 2nd one, is there some issue with you merging it?
